utility/session.go

```diff
@@ -0,0 +1,138 @@
+package utility
+
+import (
+	"encoding/binary"
+	"encoding/hex"
+	"github.com/pokt-network/pocket/shared/crypto"
+	"github.com/pokt-network/pocket/shared/modules"
+	"github.com/pokt-network/pocket/utility/types"
+)
+
+type Session interface {
+	NewSession(sessionHeight int64, blockHash string, geoZone GeoZone, relayChain RelayChain, application modules.Actor) (Session, types.Error)
+	GetServiceNodes() []modules.Actor // the ServiceNodes providing Web3 to the application
+	GetFishermen() []modules.Actor    // the Fishermen monitoring the serviceNodes
+	GetApplication() modules.Actor    // the Application consuming the web3 access
+	GetRelayChain() RelayChain        // the chain identifier of the web3
+	GetGeoZone() GeoZone              // the geolocation zone where all are registered
+	GetSessionHeight() int64          // the block height when the session started
+}
+
+type RelayChain Identifier
+type GeoZone Identifier
+
+type Identifier interface {
+	Name() string
+	ID() string
+	Bytes() []byte
+}
+
+var _ Session = &session{}
+
+type session struct {
+	serviceNodes  []modules.Actor
+	fishermen     []modules.Actor
+	application   modules.Actor
+	relayChain    RelayChain
+	geoZone       GeoZone
+	blockHash     string
+	key           []byte
+	sessionHeight int64
+}
+
+func (s *session) NewSession(sessionHeight int64, blockHash string, geoZone GeoZone, relayChain RelayChain, application modules.Actor) (session Session, err types.Error) {
+	s.sessionHeight = sessionHeight
+	s.blockHash = blockHash
+	s.geoZone = geoZone
+	s.relayChain = relayChain
+	s.application = application
+	s.key, err = s.sessionKey()
+	if err != nil {
+		return
+	}
+	s.serviceNodes = s.findClosestXServiceNodes()
+	s.fishermen = s.findClosestYFishermen()
+	return s, nil
+}
+
+// use the seed information to determine a SHA3Hash that is used to find the closest N actors based
+// by comparing the sessionKey with the actors' public key
+func (s *session) sessionKey() ([]byte, types.Error) {
```

Comment:
I think we shold put the helper functions below the official interface implementation

Reply:
Just because they are internal doesn't mean they are 'helper' 

This is the core logic of the Session protocol and is organized as such